### PR TITLE
Refactor folding.

### LINKED ISSUE #5168: Folding: zc and zo handle a count wrong
**Describe the bug**
see topic ;-)

**To Reproduce**

1.  open any document with at least 4 levels of folding, vim source will do ;-)
2.  place cursor on a line deeply nested (have all folds open now)
3.  press "2zc"
4.  this will close 4 levels of folds, not 2.

**Additional context**
The reason is that the operators run for each count, so it multiplies. I will have a PR for this.

### PR DESCRIPTION
**Pull request by sql-koala**
**What this PR does / why we need it**:
This does a refactor the folding classes a bit and fixes two bugs.

**Which issue(s) this PR fixes**

fixes #5168 
fices #5167  

**Special notes for your reviewer**:
this refactors exec to do these steps (for all derived Commands):
1. call vscode with folding command
2. sync cursor
3. set mode to Normal (orig vim does this too)

Bugs fixed:
- we need to sync cursors after zM, zC
- zo, zc must not run for each count

### CODE REVIEW

File: src/actions/commands/actions.ts [hunk 1]
```diff
@@ -1183,7 +1185,16 @@ abstract class CommandFold extends BaseCommand {
...
-    await vscode.commands.executeCommand(this.commandName);
+    let timesToRepeat = vimState.recordedState.count || 1;
+    if (this.direction !== '') {
+      await vscode.commands.executeCommand(this.commandName, {
+        levels: timesToRepeat,
+        direction: this.direction,
+      });
+    } else {
+      await vscode.commands.executeCommand(this.commandName);
+    }
```
Comment (sql-koala) [inline src/actions/commands/actions.ts:1196]: would be interested if this can be done with a neat one liner?

Comment (J-Fields) [reply src/actions/commands/actions.ts:1196]:
```typescript
const args = this.direction != '' ? { levels: timesToRepeat, direction: this.direction} : undefined;
await vscode.commands.executeCommand(this.commandName, args);
```
seems like the nicest way of writing this

File: src/actions/commands/actions.ts [hunk 2]
```diff
@@ -1171,9 +1171,11 @@ class CommandRepeatSubstitution extends BaseCommand {
   }
 }
 
+type FoldDirection = 'up' | 'down' | '';
```
Comment (J-Fields) [inline src/actions/commands/actions.ts:1174]: Any reason to use `''` over `undefined`?

File: src/actions/commands/actions.ts (hunk 1 above)
Comment (sql-koala) [reply src/actions/commands/actions.ts]: linter turned that into 4 lines :D

**Comment by J-Fields** [comment]
Thanks!

**Pull request merged**